From 5c66a672f4d268990de569f226eb0c4aec57268d Mon Sep 17 00:00:00 2001
From: Simone Tripodi <simonetripodi@apache.org>
Date: Thu, 23 Jun 2011 15:04:01 +0000
Subject: [PATCH] [DIGESTER-18] ObjectCreateRule shouldn't keep className as a
 field

this fix contains a little variant respect to the proposed suggestion: users can still create classes by specifying just the class name, but when passed the Class object in the constructor, it doesn't need to be load dynamically

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/digester/trunk@1138914 13f79535-47bb-0310-9956-ffa450edef68
---
 .../commons/digester3/ObjectCreateRule.java   | 40 ++++++++++++-------
 1 file changed, 26 insertions(+), 14 deletions(-)

diff --git a/src/main/java/org/apache/commons/digester3/ObjectCreateRule.java b/src/main/java/org/apache/commons/digester3/ObjectCreateRule.java
index 1badcd91d..d42aff9ba 100644
--- a/src/main/java/org/apache/commons/digester3/ObjectCreateRule.java
+++ b/src/main/java/org/apache/commons/digester3/ObjectCreateRule.java
@@ -49,6 +49,7 @@ public ObjectCreateRule( String className )
     public ObjectCreateRule( Class<?> clazz )
     {
         this( clazz.getName(), (String) null );
+        this.clazz = clazz;
     }
 
     /**
@@ -73,6 +74,7 @@ public ObjectCreateRule( String className, String attributeName )
     public ObjectCreateRule( String attributeName, Class<?> clazz )
     {
         this( clazz.getName(), attributeName );
+        this.clazz = clazz;
     }
 
     // ----------------------------------------------------- Instance Variables
@@ -82,6 +84,11 @@ public ObjectCreateRule( String attributeName, Class<?> clazz )
      */
     protected String attributeName = null;
 
+    /**
+     * The Java class of the object to be created.
+     */
+    protected Class<?> clazz = null;
+
     /**
      * The Java class name of the object to be created.
      */
@@ -96,24 +103,29 @@ public ObjectCreateRule( String attributeName, Class<?> clazz )
     public void begin( String namespace, String name, Attributes attributes )
         throws Exception
     {
-        // Identify the name of the class to instantiate
-        String realClassName = className;
-        if ( attributeName != null )
+        Class<?> clazz = this.clazz;
+
+        if ( clazz == null )
         {
-            String value = attributes.getValue( attributeName );
-            if ( value != null )
+            // Identify the name of the class to instantiate
+            String realClassName = className;
+            if ( attributeName != null )
             {
-                realClassName = value;
+                String value = attributes.getValue( attributeName );
+                if ( value != null )
+                {
+                    realClassName = value;
+                }
+            }
+            if ( getDigester().getLogger().isDebugEnabled() )
+            {
+                getDigester().getLogger().debug( "[ObjectCreateRule]{" + getDigester().getMatch() + "}New "
+                                                 + realClassName );
             }
-        }
-        if ( getDigester().getLogger().isDebugEnabled() )
-        {
-            getDigester().getLogger().debug( "[ObjectCreateRule]{" + getDigester().getMatch() + "}New "
-                                             + realClassName );
-        }
 
-        // Instantiate the new object and push it on the context stack
-        Class<?> clazz = getDigester().getClassLoader().loadClass( realClassName );
+            // Instantiate the new object and push it on the context stack
+            clazz = getDigester().getClassLoader().loadClass( realClassName );
+        }
         Object instance = clazz.newInstance();
         getDigester().push( instance );
     }
